fluka_sobp_source/source_sampler.f

```diff
@@ -0,0 +1,635 @@
+!!  - DE : energy spread (sigma) in GeV/amu
+!!  - X, Y : position (in cm) of the beamlet/spot center
+!!  - FWHM_X, FWHM_Y, FWHM : spot size (in cm)
+!!
+!! All above-mentioned quantities are defined at beam source location (typically nozzle exit)
+!!
+!! For more details, see SHIELD-HIT12A manual http://shieldhit.org/index.php?id=documentation
+!!
+!!
+!!
+!! -------------------- Beam configuration in the INPUT FILE -------------------------------------
+!! Beam momentum/energy, its spread and divergence is usually specified in BEAM input card.
+!! Also beam width is specified there.
+!! All these values except momentum spread will be ignored and overridden.
+!! In case 5- or 6-columns sobp.dat file format (the case where DE is missing) is used,
+!! then beam momentum will be taken from the input card, otherwise is will be ignored and overriden.
+!!
+!! Center of the beam spot is usually defined in BEAMPOS card in the input file,
+!! which contains X,Y,Z positions and direction cosines.
+!! Whatever is specified as X,Y and direction cosines will be ignored and overridden
+!! by this custom source. SDUM value in BEAMPOS card will also be ignored.
+!!
+!! In order to activate this custom source, please add SOURCE card to the input file.
+!! It has one optional parameter: WHAT(1), which (if present) will be understood
+!! as the position of the virtual source. We assume virtual source is located on
+!! negative part of Z axis, thus this position is given as single negative number.
+!! Another parameters (so called SDUM) is the filename containing table of numbers
+!! with beam specification. Typically its called sobp.dat.
+!! An example SOURCE card looks like that:
+!!
+!! SOURCE           -205.0                                                sobp.dat
+!!
+!!
+!! For more details, see FLUKA documentation:
+!!  - BEAM    card http://www.fluka.org/fluka.php?id=man_onl&sub=12
+!!  - BEAMPOS card http://www.fluka.org/fluka.php?id=man_onl&sub=14
+!!  - SOURCE  card http://www.fluka.org/fluka.php?id=man_onl&sub=71
+!!
+!! -------------------------------------------------------------------------------------------------
+
+
+!! TODO add sanity check if randomly sampled energy is positive
+
+!! ==================================================================================================
+!! ==================================================================================================
+
+!> @brief
+!! Returns number of text blocks separated with white characters (spaces)
+!! It corresponds to number of columns in CSV file.
+!! @param[in] STRING
+!! @retval LENGTH number of columns
+      INTEGER FUNCTION NOCOLS(ASTRING)
+      IMPLICIT NONE
+      CHARACTER*(*) ASTRING   ! input - string of arbitrary length
+      INTEGER I               !
+      NOCOLS = 0
+
+*     make a loop and check for non-space followed by space
+*     each occurence will mark end of a column
+      DO I = 1, LEN(ASTRING)-1
+         IF((ASTRING(I:I) .NE. ' ') .AND.
```

Comment:
for Fortran we could adapt similar style that we do for Python. IMHO readability increases adding whitespaces e.g. after `IF`.
Same for the line before, operators may be surrounded by whitespace.